File: frontend/src/services/productService.ts
import { apiClient, extractData } from './api';
import { ApiResponse, PageResponse } from '@/types/api';
import { Product, CreateProductRequest, UpdateProductRequest } from '@/types/product';

/**
 * Product service for API calls
 */
export const productService = {
  /**
   * Get all products (paginated)
   */
  getAll: async (page = 0, size = 20, active = true): Promise<PageResponse<Product>> => {
    const response = await apiClient.get<ApiResponse<PageResponse<Product>>>('/products', {
      params: { page, size, active },
    });
    return extractData(response.data);
  },

  /**
   * Get product by ID
   */
  getById: async (id: string): Promise<Product> => {
    const response = await apiClient.get<ApiResponse<Product>>(`/products/${id}`);
    return extractData(response.data);
  },

  /**
   * Get product by SKU
   */
  getBySku: async (sku: string): Promise<Product> => {
    const response = await apiClient.get<ApiResponse<Product>>(`/products/sku/${sku}`);
    return extractData(response.data);
  },

  /**
   * Get products by category
   */
  getByCategory: async (categoryId: string, page = 0, size = 20): Promise<PageResponse<Product>> => {
    const response = await apiClient.get<ApiResponse<PageResponse<Product>>>(
      `/products/category/${categoryId}`,
      { params: { page, size } }
    );
    return extractData(response.data);
  },

  /**
   * Search products
   */
  search: async (query: string, page = 0, size = 20): Promise<PageResponse<Product>> => {
    const response = await apiClient.get<ApiResponse<PageResponse<Product>>>('/products/search', {
      params: { query, page, size },
    });
    return extractData(response.data);
  },

  /**
   * Create new product
   */
  create: async (product: CreateProductRequest): Promise<Product> => {
    const response = await apiClient.post<ApiResponse<Product>>('/products', product);
    return extractData(response.data);
  },

  /**
   * Update product
   */
  update: async (id: string, product: UpdateProductRequest): Promise<Product> => {
    const response = await apiClient.put<ApiResponse<Product>>(`/products/${id}`, product);
    return extractData(response.data);
  },

  /**
   * Delete product
   */
  delete: async (id: string): Promise<void> => {
    await apiClient.delete(`/products/${id}`);
  },

  /**
   * Update product stock
   */
  updateStock: async (id: string, quantity: number): Promise<Product> => {
    const response = await apiClient.patch<ApiResponse<Product>>(`/products/${id}/stock`, null, {
      params: { quantity },
    });
    return extractData(response.data);
  },

  /**
   * Activate product
   */
  activate: async (id: string): Promise<Product> => {
    const response = await apiClient.patch<ApiResponse<Product>>(`/products/${id}/activate`);
    return extractData(response.data);
  },

  /**
   * Deactivate product
   */
  deactivate: async (id: string): Promise<Product> => {
    const response = await apiClient.patch<ApiResponse<Product>>(`/products/${id}/deactivate`);
    return extractData(response.data);
  },
};

File: backend/product-service/src/main/java/com/dukkan/product/service/ProductService.java
package com.dukkan.product.service;

import com.dukkan.product.dto.ProductDTO;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;
import java.util.List;
import java.util.UUID;

/**
 * Service interface for Product business logic.
 * Following Interface Segregation and Dependency Inversion Principles.
 */
public interface ProductService {

    /**
     * Create a new product
     */
    ProductDTO createProduct(ProductDTO productDTO);

    /**
     * Get product by ID
     */
    ProductDTO getProductById(UUID id);

    /**
     * Get product by SKU
     */
    ProductDTO getProductBySku(String sku);

    /**
     * Get all products (paginated)
     */
    Page<ProductDTO> getAllProducts(Pageable pageable);

    /**
     * Get all active products (paginated)
     */
    Page<ProductDTO> getActiveProducts(Pageable pageable);

    /**
     * Get products by category (paginated)
     */
    Page<ProductDTO> getProductsByCategory(UUID categoryId, Pageable pageable);

    /**
     * Search products by name or description
     */
    Page<ProductDTO> searchProducts(String searchTerm, Pageable pageable);

    /**
     * Get low stock products
     */
    List<ProductDTO> getLowStockProducts(Integer threshold);

    /**
     * Get out of stock products
     */
    List<ProductDTO> getOutOfStockProducts();

    /**
     * Update an existing product
     */
    ProductDTO updateProduct(UUID id, ProductDTO productDTO);

    /**
     * Delete a product
     */
    void deleteProduct(UUID id);

    /**
     * Update product stock
     */
    ProductDTO updateStock(UUID id, Integer quantity);

    /**
     * Reduce product stock (for orders)
     */
    ProductDTO reduceStock(UUID id, Integer quantity);

    /**
     * Increase product stock (for restocking)
     */
    ProductDTO increaseStock(UUID id, Integer quantity);

    /**
     * Activate a product
     */
    ProductDTO activateProduct(UUID id);

    /**
     * Deactivate a product
     */
    ProductDTO deactivateProduct(UUID id);

    /**
     * Upload an image for a product
     *
     * @param productId Product ID
     * @param file Image file to upload
     * @return Updated product DTO with new image URL
     * @throws IOException if file upload fails
     */
    ProductDTO uploadProductImage(UUID productId, MultipartFile file) throws IOException;

    /**
     * Remove an image from a product
     *
     * @param productId Product ID
     * @param imageUrl Image URL to remove
     * @return Updated product DTO without the removed image
     * @throws IOException if file deletion fails
     */
    ProductDTO removeProductImage(UUID productId, String imageUrl) throws IOException;
}
